language/tools/move-package/src/compilation/compiled_package.rs

```diff
@@ -317,39 +336,68 @@ impl CompiledPackage {
     pub fn transitive_compiled_modules(&self) -> Modules {
         Modules::new(
             self.transitive_dependencies()
-                .iter()
                 .flat_map(|compiled_package| &compiled_package.compiled_units)
-                .chain(self.compiled_units.iter())
-                .filter_map(|unit| match unit {
+                .filter_map(|unit| match &unit.unit {
                     CompiledUnit::Module(NamedCompiledModule { module, .. }) => Some(module),
                     CompiledUnit::Script(_) => None,
                 }),
         )
     }
 
-    /// Returns `CompiledPackage`s in dependency order and deduped
-    pub fn transitive_dependencies(&self) -> &[CompiledPackage] {
-        &self.dependencies
+    /// Returns `CompiledPackage`s deduped
+    pub fn transitive_dependencies(&self) -> impl Iterator<Item = &CompiledPackage> {
+        self.dependencies.iter().chain(vec![self])
+    }
+
+    pub fn modules(&self) -> impl Iterator<Item = &CompiledUnitWithSource> {
+        self.compiled_units
+            .iter()
+            .filter(|unit| matches!(unit.unit, CompiledUnit::Module(_)))
+    }
+
+    pub fn scripts(&self) -> impl Iterator<Item = &CompiledUnitWithSource> {
+        self.compiled_units
+            .iter()
+            .filter(|unit| matches!(unit.unit, CompiledUnit::Script(_)))
+    }
+
+    fn can_load_cached(
+        package: &OnDiskCompiledPackage,
+        resolution_graph: &ResolvedGraph,
+        resolved_package: &ResolvedPackage,
+        is_root_package: bool,
+    ) -> bool {
+        !(package.has_source_changed_since_last_compile(resolved_package)
+                || package.are_build_flags_different(&resolution_graph.build_options)
+                // force root package recompilation in test mode
+                || resolution_graph.build_options.test_mode && is_root_package
+                || resolution_graph.build_options.force_recompilation) &&
+                // Dive deeper to make sure that instantiations haven't changed since that
+                // can be changed by other packages above us in the dependency graph possibly
+                package.package.compiled_package_info.address_alias_instantiation
+                    == resolved_package.resolution_table
     }
 
     pub(crate) fn build<W: Write>(
         w: &mut W,
         project_root: &Path,
         resolved_package: ResolvedPackage,
-        dependencies: Vec<CompiledPackage>,
+        dependencies: Vec<(CompiledPackage, bool)>,
         resolution_graph: &ResolvedGraph,
         is_root_package: bool,
         mut compiler_driver: impl FnMut(
             Compiler,
             bool,
         )
             -> Result<(FilesSourceText, Vec<AnnotatedCompiledUnit>)>,
-    ) -> Result<CompiledPackage> {
+    ) -> Result<(CompiledPackage, bool)> {
```

Comment:
I sometimes find boolean result values difficult to grok; maybe this could use a comment explaining what `true`/`false` means semantically? Or maybe an enum that names the `true` and the `false` cases?

Based on the code below I'm thinking that `true` means "the package and its dependencies are all in the cache" and `false` means "we needed to recompile the package and its dependencies?"

Reply:
Correct! I think making these an enum would be nice/more transparent. Will update